pkg/sql/logictest/tests/5node/5node_test: TestLogic_distsql_stats failed

pkg/sql/logictest/tests/5node/5node_test.TestLogic_distsql_stats [failed](https://tanzanite.cluster.engflow.com/invocations/default/4dba7238-a82b-442b-bb99-2560d9537b36?testReportRun=4&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cvc3FsL2xvZ2ljdGVzdC90ZXN0cy81bm9kZTo1bm9kZV90ZXN0) on release-24.3 @ [5be83d7c3461add218e04a15e158139ef945ea6c](https://github.com/cockroachdb/cockroach/commits/5be83d7c3461add218e04a15e158139ef945ea6c):


```
    logic.go:3107: let $json_stats = [{"avg_size": 4, "columns": ["x"], "created_at": "2025-02-25 10:07:40.686857", "distinct_count": 3, "histo_buckets": [{"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "hello"}, {"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "howdy"}, {"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "hi"}], "histo_col_type": "test.public.e", "histo_version": 3, "id": 1049911822550073349, "name": "s", "null_count": 0, "row_count": 3}, {"avg_size": 3, "columns": ["y"], "created_at": "2025-02-25 10:07:40.686857", "distinct_count": 3, "histo_buckets": [{"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "hello"}, {"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "howdy"}, {"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "hi"}], "histo_col_type": "test.public.e", "histo_version": 3, "id": 1049911822561509381, "name": "s", "null_count": 0, "row_count": 3}, {"avg_size": 7, "columns": ["z"], "created_at": "2025-02-25 10:07:40.686857", "distinct_count": 3, "histo_buckets": [{"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "ARRAY['hello':::test.public.e]"}, {"distinct_range": 1, "num_eq": 1, "num_range": 1, "upper_bound": "ARRAY['hi':::test.public.e]"}], "histo_col_type": "test.public.e[]", "histo_version": 3, "id": 1049911822573797381, "name": "s", "null_count": 0, "row_count": 3}]
    logic.go:3107: let $hist_id_1 = 1049911823609757701
    logic.go:3107: let $hist_id_1 = 1049911823893725189
    logic.go:3107: let $hist_id_1 = 1049911824286089221
    logic.go:3107: let $hist_id_1 = 1049911825088184325
[10:07:41] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 179 statements
    logic.go:3107: let $hist_id_1 = 1049911829132214277
    logic.go:3107: let $hist_id_1 = 1049911829726330885
    logic.go:3107: let $stats = [{"avg_size": 4, "columns": ["x"], "created_at": "2025-02-25 10:07:43.312118", "distinct_count": 1, "histo_buckets": [{"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "hi"}], "histo_col_type": "test.public.greeting", "histo_version": 3, "id": 1049911831150690309, "name": "s", "null_count": 0, "row_count": 1}]
    logic.go:3107: let $db = test
[10:07:43] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 228 statements
    logic.go:3107: let $hist_id_1 = 1049911833622052869
    logic.go:3107: let $hist_id_1 = 1049911833849331717
    logic.go:3107: let $hist_id_1 = 1049911833859817477
[10:07:45] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 276 statements
    logic.go:3107: let $hist_abcd_a_partial = 1049911839857344517
    logic.go:3107: let $hist_abcd_c_partial = 1049911839992512517
    logic.go:3107: let $hist_xy_x_partial = 1049911840138559493
    logic.go:3107: let $statistics_id = 1049911839711657989
    logic.go:3107: let $hist_a_null_stat_partial = 1049911840818888709
    logic.go:3107: let $hist_d_desc = 1049911841209810949
    logic.go:3107: let $hist_id_1 = 1049911844322017285
[10:07:47] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 355 statements
    logic.go:3107: let $hist_d = 1049911846657490949
    logic.go:3107: let $hist_e = 1049911846821527557
    logic.go:3107: let $hist_f = 1049911846996279301
    logic.go:3107: let $hist_crdb_internal_idx_expr = 1049911848322760709
    logic.go:3107: let $hist_d = 1049911848525266949
    logic.go:3107: let $hist_e = 1049911848681897989
    logic.go:3107: let $hist_crdb_internal_idx_expr = 1049911848843411461
    logic.go:3107: let $hist_id_int_outer_buckets_full = 1049911849477668869
    logic.go:3107: let $hist_id_timestamp_outer_buckets_full = 1049911850448715781
[10:08:03] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 425 statements
    logic.go:3107: let $job = 1049911851842666497
[10:08:29] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 430 statements
    logic.go:3107: let $hist_status = 1049911983507832834
[10:08:46] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 454 statements
    logic.go:3107: let $job = 1049911984075735041
[10:08:53] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 458 statements
[10:08:53] --- done: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats with config 5node: 462 tests, 0 failures
    logic.go:4247: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats:3600: error while processing
    logic.go:4247: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats:3600: 
        expected success, but found
        (55000) column a does not have a prior statistic
        distsql_plan_stats.go:375: in createPartialStatsPlan()
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestLogic_distsql_stats558396296
--- FAIL: TestLogic_distsql_stats (81.85s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #140353 pkg/sql/logictest/tests/5node/5node_test: TestLogic_distsql_stats failed [C-test-failure O-robot T-sql-queries branch-release-24.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_distsql_stats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48106

--- a/pkg/sql/logictest/testdata/logic_test/distsql_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_stats
@@ -2424,6 +2424,11 @@ INSERT INTO abcd VALUES
 statement ok
 INSERT INTO xy VALUES (-1, 9), (-2, 8), (5, 15), (6, 16)
 
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 statement error pgcode 0A000 creating partial statistics with a WHERE clause is not yet supported
 CREATE STATISTICS abcd_a_partial ON a FROM abcd WHERE a > 1;
 
@@ -2539,13 +2544,13 @@ INSERT INTO a_null VALUES (NULL), (1), (2);
 statement ok
 CREATE STATISTICS a_null_stat ON a FROM a_null;
 
+statement ok
+INSERT INTO a_null VALUES (NULL), (NULL), (NULL);
+
 # Clear the stat cache so that creating partial statistics has access to the
 # latest full statistic.
 statement ok
 SELECT crdb_internal.clear_table_stats_cache();
-
-statement ok
-INSERT INTO a_null VALUES (NULL), (NULL), (NULL);
 
 statement ok
 CREATE STATISTICS a_null_stat_partial ON a FROM a_null USING EXTREMES;
@@ -2577,13 +2582,13 @@ INSERT INTO d_desc VALUES (1, 10), (2, 20), (3, 30), (4, 40);
 statement ok
 CREATE STATISTICS sd ON a FROM d_desc;
 
+statement ok
+INSERT INTO d_desc VALUES (0, 0), (5, 50);
+
 # Clear the stat cache so that creating partial statistics has access to the
 # latest full statistic.
 statement ok
 SELECT crdb_internal.clear_table_stats_cache();
-
-statement ok
-INSERT INTO d_desc VALUES (0, 0), (5, 50);
 
 statement ok
 CREATE STATISTICS sdp ON a FROM d_desc USING EXTREMES;
@@ -3016,6 +3021,11 @@ ANALYZE t130817;
 statement ok
 INSERT INTO t68254 (a, b, c) VALUES (5, '5', '{"foo": {"bar": {"baz": 5}}}')
 
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 statement ok
 CREATE STATISTICS j6 ON d FROM t68254 USING EXTREMES
 
@@ -3081,10 +3091,12 @@ ALTER TABLE int_outer_buckets INJECT STATISTICS '[
 ]'
 
 statement ok
-SELECT crdb_internal.clear_table_stats_cache();
-
-statement ok
 INSERT INTO int_outer_buckets SELECT generate_series(-10, -1) UNION ALL SELECT generate_series(10000, 10009);
+
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 CREATE STATISTICS int_outer_buckets_partial ON a FROM int_outer_buckets USING EXTREMES;
@@ -3143,11 +3155,6 @@ INSERT INTO timestamp_outer_buckets VALUES
 statement ok
 CREATE STATISTICS timestamp_outer_buckets_full ON a FROM timestamp_outer_buckets;
 
-# Clear the stat cache so that creating partial statistics has access to the
-# latest full statistic.
-statement ok
-SELECT crdb_internal.clear_table_stats_cache();
-
 let $hist_id_timestamp_outer_buckets_full
 SELECT histogram_id FROM [SHOW STATISTICS FOR TABLE timestamp_outer_buckets] WHERE statistics_name = 'timestamp_outer_buckets_full'
 
@@ -3161,6 +3168,11 @@ statement ok
 INSERT INTO timestamp_outer_buckets VALUES
   ('2024-06-26 00:00:00'),
   ('2024-06-27 03:30:00');
+
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 CREATE STATISTICS timestamp_outer_buckets_partial ON a FROM timestamp_outer_buckets USING EXTREMES;
@@ -3232,6 +3244,11 @@ ALTER TABLE timestamp_outer_buckets INJECT STATISTICS '[
 
 statement ok
 INSERT INTO timestamp_outer_buckets VALUES ('2024-06-28 01:00:00');
+
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 CREATE STATISTICS timestamp_outer_buckets_partial ON a FROM timestamp_outer_buckets USING EXTREMES;
@@ -3614,6 +3631,11 @@ INSERT INTO pstat_allindex VALUES
   (5, 5, 5, 5, 5, 5,'{"5": "5"}'),
   (6, 6, 6, 6, 6, 6, '{"6": "6"}'),
   (7, 7, 7, 7, 7, 7, '{"7": "7"}');
+
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 CREATE STATISTICS pstat_allindex_partial FROM pstat_allindex USING EXTREMES;

